codec/src/types/set.rs

```diff
@@ -0,0 +1,504 @@
+    }
+}
+
+impl<K: Ord + Hash + Eq + EncodeSize> EncodeSize for BTreeSet<K> {
+    fn encode_size(&self) -> usize {
+        let mut size = self.len().encode_size();
+        for item in self {
+            size += item.encode_size();
+        }
+        size
+    }
+}
+
+impl<R: RangeConfig, Cfg: Config, K: Read<Cfg> + Clone + Ord + Hash + Eq> Read<(R, Cfg)>
+    for BTreeSet<K>
+{
+    fn read_cfg(buf: &mut impl Buf, (range, cfg): &(R, Cfg)) -> Result<Self, Error> {
+        // Read and validate the length prefix
+        let len = usize::read_cfg(buf, range)?;
+        let mut set = BTreeSet::new(); // BTreeSet does not have a capacity method
+
+        // Keep track of the last item read
+        let mut last: Option<K> = None;
+
+        // Read each item
+        for _ in 0..len {
+            let item = K::read_cfg(buf, cfg)?;
+
+            // Check if items are in ascending order
+            if let Some(ref last) = last {
+                match item.cmp(last) {
+                    Ordering::Equal => return Err(Error::Invalid("BTreeSet", "Duplicate item")),
+                    Ordering::Less => return Err(Error::Invalid("BTreeSet", "Items must ascend")),
+                    _ => {}
+                }
+            }
+            last = Some(item.clone());
+            set.insert(item);
+        }
+
+        Ok(set)
+    }
+}
+
+// ---------- HashSet ----------
+
+impl<K: Ord + Hash + Eq + Write> Write for HashSet<K> {
+    fn write(&self, buf: &mut impl BufMut) {
+        self.len().write(buf);
+
+        // Sort the items to ensure deterministic encoding
+        let mut items: Vec<_> = self.iter().collect();
+        items.sort();
+        for item in items {
+            item.write(buf);
+        }
+    }
+}
+
+impl<K: Ord + Hash + Eq + EncodeSize> EncodeSize for HashSet<K> {
+    fn encode_size(&self) -> usize {
+        let mut size = self.len().encode_size();
+        for item in self {
+            size += item.encode_size();
+        }
+        size
+    }
+}
+
+impl<R: RangeConfig, Cfg: Config, K: Read<Cfg> + Clone + Ord + Hash + Eq> Read<(R, Cfg)>
+    for HashSet<K>
+{
+    fn read_cfg(buf: &mut impl Buf, (range, cfg): &(R, Cfg)) -> Result<Self, Error> {
+        // Read and validate the length prefix
+        let len = usize::read_cfg(buf, range)?;
+        let mut set = HashSet::with_capacity(len);
+
+        // Keep track of the last item read
+        let mut last: Option<K> = None;
+
+        // Read each item
+        for _ in 0..len {
+            let item = K::read_cfg(buf, cfg)?;
+
+            // Check if items are in ascending order
+            if let Some(ref last) = last {
+                match item.cmp(last) {
+                    Ordering::Equal => return Err(Error::Invalid("HashSet", "Duplicate item")),
+                    Ordering::Less => return Err(Error::Invalid("HashSet", "Items must ascend")),
+                    _ => {}
+                }
+            }
+            last = Some(item.clone());
```

Comment:
I wonder if we can just add this after the final iteration to avoid the clone?

Reply:
Sorry didn't understand your meaning here. Did you mean that we should skip the clone on just the last iteration as an optimization (since we don't have to check `last` again)? Or something else

Reply:
Maybe easier to understand in code: https://github.com/commonwarexyz/monorepo/pull/1008